### PR TITLE
Time parse fractional days

### PR DESCRIPTION
**Pull request by evertrol**
PR to aid in the discussion at #6476 

For now, this is a suggestion, and not (yet) to be merged.

### CODE REVIEW

**Comment by evertrol** [comment]
I guess the Travis failure is unrelated? "test_skip_remote_data_astropy" failed.

**Comment by taldcroft** [comment]
Looking at the code reminded me there is an annoyance here, namely days with a leap second where you have 86401 seconds in the day for the UTC scale.  I think what you want to do is defer to ERFA instead of doing this math yourself, at least for scale = `utc`.  Something along the lines of below (hacked from the `decimalyear` format, but you can find other examples, also look at the base `TimeString`):
```
jd1, jd2 = erfa.dtf2d(scale, *vals)  # without the fraction
scale = self.scale.upper().encode('ascii')
# I forget the details of d2dtf and what to use for the `precision` arg, which is
# set to 0 here.
iyear, imon, iday, ihmsfs = erfa.d2dtf(scale, 0,  jd1, jd2 + fraction)
vals = ...
```

**Comment by mhvk** [comment]
@evertrol - I like having an actual addition, but as noted by @taldcroft this doesn't work for days with leap seconds. To see how to do this generically, have a look at `TimeDecimalYear` - essentially, you calculate the time for the integer day before and then use the fraction to interpolate, which seems foolproof.

It does imply the fractional day needs a separate path.

Separately, I'm a bit uncomfortable with this being a "hack" - could we make this a separate `in_subfmt` in which `day` has a fraction? (haven't checked whether this is easy to do...).

**Comment by taldcroft** [comment]
> Separately, I'm a bit uncomfortable with this being a "hack" - could we make this a separate in_subfmt in which day has a fraction? (haven't checked whether this is easy to do...).

I do not think this is easy without entirely punting on `strptime`.  The inexplicable lack of support for fractional seconds (or fractional anything) is what led to the original hack.  So this would morph this tiny corner case change into a fairly big change.  But of course not impossible as demonstrated in `TimeFITS`.

I can't remember now if `strptime` is faster and that was a driver.  I vaguely recall being surprised that `strptime` was slow, and there is http://joshuakugler.com/archives/27-Dont-use-strptime,-use-a-regex.html.  Would be worth checking for Py3.

**Comment by mhvk** [comment]
Looking back at `formats.py` reminded me that we already turn `strptime` formats into `regex` by `_regexify_subfmts`... It does seem one could add a `fracdate` subfmt and insist that `date` ends with an integer.

**Comment by evertrol** [comment]
@taldcroft `time.strptime` is about 4 times slower than a regex for an ISO time, for both Python 2 and 3.
But `time.strptime` will test for invalid months, days, hours, minutes or seconds (including day=31 for even months). It also allows for leap seconds. That likely explains the slowness.

Unless these checks are performed elsewhere in the `astropy.time` module (I haven't digged that deep into the module yet), it may be considered worth to have these checks. Then again, the `TimeFITS` sub-formats suggest that using a regex is fine, and I presume checks for incorrect values and leap seconds are indeed performed elsewhere.

**Comment by evertrol** [comment]
@mhvk Do you suggest to add a separate `fracdate` subformat (and possibly a `yfracday` subformat as well)? And then force `date` and `yday` to have no digits?

Is there any reason to not allow a more generic `date` and `yday` that optionally allow for fractional days, provided we can change the `strptime` format strings all to regexes?
If that would be the case, I would then also change `..._hms` formats to use regexes and allow fractional seconds.

The only reason for having separate subformats for fractional days, is to let the user make a conscious decision about what to use. Perhaps there is technical reason not to do this that I miss?
It's already implicit for fractional seconds: the user has no say there whether to allow them or not (which, of course, led to this PR and corresponding issue in the first place).

**Comment by mhvk** [comment]
@evertrol - the explicit subformat would indeed also imply ensuring that in the regular formats `day` is an integer. It would not matter on normal input - the code checks all possible sub-formats. But would make it possible to use fractional days also on output.

**Comment by mhvk** [comment]
As for `strptime` checking of correct days, etc., that is indeed irrelevant, as that is redone by the `erfa.dtf2d` call.

**Comment by evertrol** [comment]
This is an update / replacement of my previous commits to implement fractional days parsing for `astropy.time.Time`.

It's still very much a hack, but other than completely rewriting half the `formats.py` module, I don't readily see an easy way to otherwise support this.

I've moved the fractional second part handling into the loop of the `parse_string` routine, and similar for fraction day handling. Fractional seconds are easy to pass on to the output, since the `vals` list (and related routines) accept seconds in double format. 

For fractional days, this is not so easy. Thus, the fraction is first added to the Julian day for the date (at midnight, since hms will all be 0 for a fractional day date), with the JD calculcated from the year-month-day using the `erfa.dtf2d` routine; and then things are converted back to the proper hms using the `erfa.d2dtf` routine. The back-and-forth is done to prevent issues with leap-seconds, as suggested above by taldcroft(*).
This is obviously not very efficient, but seemed the easiest to achieve (and I'm not too worried about efficiency when parsing input strings as time).

For this to work, I've replaced various `strptime` formats with regexes that accept fractional days or seconds.

I have not worked on the output format, but as far as I can tell, times will be formatted using the relevant ISO or yday formatting (that is, the fractional day itself will disappear and be replaced by HH:MM:SS.SSS).

The PR itself is still incomplete, with missing documentation updates, doc-string updates and changes  update. For now, I'm mostly concerned with overlooked edge cases (I've only added a few new tests for basic behaviour testing).


(*) My understanding is that `d1` in the `d2dtf`/`dtf2d` routines is always the JD at noon (integer + 0.5), and `d2` always the fraction, but this isn't 100% clear to me from the comments of [the relevant C functions](https://github.com/astropy/astropy/blob/master/cextern/erfa/dtf2d.c), e.g. point 3: "where normally d1 is the Julian Day Number". From a few tests, it seems this is still the case for leap seconds (point 4), so I'm unclear when is it not the JD number.

File: astropy/time/formats.py [hunk 1]
```diff
@@ -720,11 +712,33 @@ def parse_string(self, timestr, subfmts):
                 if tm is None:
                     continue
                 tm = tm.groupdict()
+                if '.' in tm.get('sec', ''):
```
Comment (taldcroft) [inline astropy/time/formats.py]: I wonder if you can tidy this up a bit by making the fractional part a named component.  E.g.
```
In [24]: r = r'(?P<sec>\d{1,2})(?P<fracsec>\.\d*)?'

In [25]: m = re.match(r, '12.1')

In [26]: m.groupdict()
Out[26]: {'fracsec': '.1', 'sec': '12'}

In [27]: m = re.match(r, '12.')

In [28]: m.groupdict()
Out[28]: {'fracsec': '.', 'sec': '12'}

In [29]: m = re.match(r, '12')

In [30]: m.groupdict()
Out[30]: {'fracsec': None, 'sec': '12'}
```
Then:
```
fracsec = tm.get('fracsec')
if fracsec is not None:
    fracsec = float('0' + fracsec)
```

**Comment by taldcroft** [comment]
@evertrol @mhvk - Please see the alternative implementation here:
 https://github.com/astropy/astropy/compare/master...taldcroft:time-frac-days?expand=1

The idea is to allow for generic chopping of a fractional bit up front which might be a day or it might be a sec.  This is determined by a class attribute that specifically calls out sub formats that end with either day or sec.  The fractional second is added within `parse_string` while the fractional day is added to `jd2` outside of `parse_string`.

My POC is not at all complete but intended as an idea for @evertrol to finish.  Thoughts?
```
In [2]: t = Time('2001-01-01')

In [3]: t
Out[3]: <Time object: scale='utc' format='iso' value=2001-01-01 00:00:00.000>

In [4]: t = Time('2001-01-01.5')

In [5]: t
Out[5]: <Time object: scale='utc' format='iso' value=2001-01-01 12:00:00.000>

In [6]: t = Time('2001-01-01 12:00:00.5')

In [7]: t
Out[7]: <Time object: scale='utc' format='iso' value=2001-01-01 12:00:00.500>
```

**Comment by mhvk** [comment]
I think the implementation of @taldcroft is cleaner, but I liked in @evertrol's that the fractional day has an explicit subformat name, so that, if one wants to be strict and not allow that format (it is not part of the ISO standard, after all), it is possible to do that (I think, with a suitable choice of subfmt; correct?).

**Comment by pllim** [comment]
Since this is experimental, I am removing the milestone. We have the bot to ping us anyway when this gets too old.

**Pull request closed without merging**